scala/e2e-tests/src/test/scala/uk/gov/nationalarchives/e2etests/StepDefs.scala

```diff
@@ -0,0 +1,193 @@
+import software.amazon.awssdk.transfer.s3.model.CompletedUpload
+import uk.gov.nationalarchives.DADynamoDBClient.DADynamoDbWriteItemRequest
+import uk.gov.nationalarchives.e2etests.StepDefs.{Config, *}
+import uk.gov.nationalarchives.{DADynamoDBClient, DAS3Client, DASFNClient, DASQSClient}
+
+import java.net.URI
+import scala.concurrent.duration.*
+import java.time.{Instant, LocalDate}
+import java.util.UUID
+import scala.reflect.ClassTag
+import scala.util.Random
+
+class StepDefs[F[_]: Async](messageUuidCell: AtomicCell[F, List[SqsMessage]]) {
+
+  val s3Client: DAS3Client[F] = DAS3Client[F]()
+  val sqsClient: DASQSClient[F] = DASQSClient[F]()
+  val dynamoClient: DADynamoDBClient[F] = DADynamoDBClient[F]()
+  val sfnClient: DASFNClient[F] = DASFNClient[F]()
+  val config: Config = ConfigSource.default.loadOrThrow[Config]
+  val ingestComplete = "preserve.digital.asset.ingest.complete"
+  val ingestUpdate = "preserve.digital.asset.ingest.update"
+  val ccDiskStatus = "Asset has been written to custodial copy disk."
+  val assetErrorStatus = "There has been an error ingesting the asset."
+
+  given SelfAwareStructuredLogger[F] = Slf4jFactory.create[F].getLogger
+
+  private def toDynamoString(value: String): AttributeValue = AttributeValue.builder.s(value).build
+
+  def createBatch(ref: Ref[F, List[UUID]]): F[Unit] = ref.get.flatMap { ids =>
+    val groupId = s"TST_${UUID.randomUUID}"
+    val batchId = s"${groupId}_0"
+    Logger[F].info(s"Creating batch $groupId with ${ids.size} ids") >>
+    ids.traverse { id =>
+      val input = AggregatorInputMessage(id, URI.create(s"s3://${config.inputBucket}/$id"))
+      val writeItem = Map("assetId" -> toDynamoString(id.toString), "groupId" -> toDynamoString(groupId), "message" -> toDynamoString(input.asJson.noSpaces))
+      dynamoClient.writeItem(DADynamoDbWriteItemRequest(config.lockTableName, writeItem, Some(s"attribute_not_exists(assetId)")))
+    } >> sfnClient.startExecution(config.preingestSfnArn, SFNArguments(groupId, batchId, 0), batchId.some).void
+  }
+
+  def sendMessages(ref: Ref[F, List[UUID]]): F[Unit] = ref.get.flatMap { ids =>
+    val sqsClient = DASQSClient[F]()
+    val send = sqsClient.sendMessage[SqsInputMessage](config.ingestSqsQueue)(_, None, 0)
+    ids.traverse { id =>
+      send(SqsInputMessage(id, config.inputBucket))
+    }.void
+  }
+
+  def createFile(bucket: String, key: String, bytes: Array[Byte]): F[CompletedUpload] =
+    Stream
+      .emits[F, Byte](bytes)
+      .chunks
+      .map(_.toByteBuffer)
+      .toPublisherResource.use { pub =>
+        s3Client.upload(bucket, key, FlowAdapters.toPublisher(pub))
+      }
+
+  def makeReference(length: Int): String =
+    Random.shuffle(('A' to 'Z') ++ ('0' to '9')).take(length).mkString
+
+  def createMetadataJson(id: UUID, checksum: String, invalidMetadata: Boolean): Array[Byte] =
+    def generateValue[T](ifPresent: T) = if invalidMetadata && Random.nextInt(100) > 50 then None else ifPresent.some
```

Comment:
1. Maybe `ifValuePresent` instead of `ifPresent`?
2. Even though what you have is fine, you could use `Random.nextBoolean()`.
3. Wait, why do we need `&& Random.nextInt(100) > 50`? Wouldn't that mean that if `invalidMetadata` is true, it will only return invalid metadata 50% of the time?
4. As well as `None,` should we add more values that could fail so `Random.shuffle(List(None, Some("NoSpace"), Some(""))).head`?

Reply:
1. There's value in the method name, I don't thinkit needs duplicating here.
2. `nextBoolean` is a good idea, I didn't know that.
3. It always has an empty Series for the invalid metadata because as you say, half the time it would be valid. It just randomly assigns other things to be empty as well. 
4. I've added that in for the series because that's where most of the validation is anyway.